src/closure.rs

```diff
@@ -0,0 +1,163 @@
+// TODO: support marshaller.
+
+use std::mem;
+use std::ptr;
+
+use libc::{c_uint, c_void};
+
+use gobject_ffi;
+use translate::{FromGlibPtrFull, FromGlibPtrNone, Stash, StashMut, ToGlibPtr, ToGlibPtrMut, Uninitialized};
+use Value;
+use ToValue;
+
+pub struct Closure {
+    closure: *mut gobject_ffi::GClosure,
+}
+
+impl Closure {
+    pub fn new<F: Fn(&[Value]) -> Option<Value> + 'static>(callback: F) -> Self {
+        unsafe extern "C" fn marshal<F: Fn(&[Value]) -> Option<Value> + 'static>(_closure: *mut gobject_ffi::GClosure,
+            return_value: *mut gobject_ffi::GValue, n_param_values: c_uint, mut param_values: *mut gobject_ffi::GValue,
+            _invocation_hint: *mut c_void, marshal_data: *mut c_void)
+        {
+            let mut values = vec![];
+            for _ in 0 .. n_param_values {
+                let value: Value = FromGlibPtrNone::from_glib_none(param_values);
+                values.push(value);
+                param_values = param_values.offset(1);
+            }
+            let callback: &Box<F> = mem::transmute(marshal_data as *mut _);
+            let result = callback(&values);
+            match result {
+                Some(result) => *return_value = result.into_raw(),
+                None => {
+                    let result = Value::uninitialized();
+                    let mut gvalue = result.into_raw();
+                    gvalue.g_type = gobject_ffi::G_TYPE_NONE;
+                    *return_value = gvalue;
+                },
+            }
+            // TODO: remove this.
+            for value in values {
+                mem::forget(value);
+            }
+        }
+
+        unsafe extern "C" fn finalize<F: Fn(&[Value]) -> Option<Value> + 'static>(notify_data: *mut c_void,
+            _closure: *mut gobject_ffi::GClosure)
+        {
+            let _callback: Box<Box<F>> = Box::from_raw(notify_data as *mut _);
+            // callback is dropped here.
+        }
+
+        unsafe {
+            let size = 4 + 4 + 3 * mem::size_of::<*mut c_void>() as u32;
+            let closure = gobject_ffi::g_closure_new_simple(size, ptr::null_mut());
+            assert_ne!(closure, ptr::null_mut());
+            let callback = Box::new(Box::new(callback));
```

Comment:
I followed the style used by the code generated by `gir`, for instance in [`gtk::button`](https://github.com/gtk-rs/gtk/blob/master/src/auto/button.rs#L294).
I don't know the reason to double box: can @GuillaumeGomez or @EPashkin explain, please?